akd_audit_quorum/src/comms/nonces.rs

```diff
@@ -0,0 +1,102 @@
+//
+// This source code is licensed under both the MIT license found in the
+// LICENSE-MIT file in the root directory of this source tree and the Apache
+// License, Version 2.0 found in the LICENSE-APACHE file in the root directory
+// of this source tree.
+
+//! This module provides management for message nonces (incoming & outgoing).
+//!
+//! The purpose of managing nonces is to protect against replay attacks in inter-node
+//! messages. The [`NonceManager`] manages both _expected_ incoming nonces from nodes
+//! and the outgoing nonces for messages being sent to nodes.
+
+use super::{NodeId, Nonce};
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+/// Manages the state of incoming & outgoing "nonces" per-node,
+/// which are effectively message counts to/from each node
+pub struct NonceManager {
+    outgoing: Arc<RwLock<HashMap<NodeId, Nonce>>>,
+    incoming: Arc<RwLock<HashMap<NodeId, Nonce>>>,
+}
+
+impl Default for NonceManager {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl NonceManager {
+    /// Create a new nonce manager instance
+    pub fn new() -> Self {
+        Self {
+            outgoing: Arc::new(RwLock::new(HashMap::new())),
+            incoming: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Retrieve the "next" nonce for message send, will auto-increment the nonce
+    /// mapping
+    pub async fn get_next_outgoing_nonce(&self, to: NodeId) -> Nonce {
+        let mut guard = self.outgoing.write().await;
+        let next = guard.get(&to).map_or(0, |a| *a + 1);
+        // safe the updated nonce
+        guard.insert(to, next);
+        next
+    }
+
+    /// Validate the incoming nonce from the specified node
+    pub async fn validate_nonce(
+        &self,
+        from: NodeId,
+        nonce: Nonce,
+    ) -> Result<(), crate::comms::CommunicationError> {
+        if from == u64::MAX {
+            // special case for new-node test's
+            return Ok(());
+        }
+
+        let expected = self.get_expected_nonce(from).await;
+
+        if nonce == expected {
+            self.incoming.write().await.insert(from, expected + 1);
+            Ok(())
+        } else {
+            let err = crate::comms::CommunicationError::NonceError(
+                from,
+                nonce,
+                format!(
+                    "Nonce mismatch in raft inter-messages: Node {}, Nonce: {}, Expected Nonce: {}",
+                    from, nonce, expected
+                ),
+            );
+
+            // reject this message but additionaly reset the nonce to what was received + 1. This handles when
+            // nodes restart or get corrupted and nonces may legitimately drift apart. In the event
+            // that a reply attack occurs, we're rejecting the message anyways if the nonce doesn't match
+            // what's expected so it's still protected.
+            self.incoming.write().await.insert(from, nonce + 1);
```

Comment:
Wouldn't that allow replay attacks repeating any two consecutive requests from any node? Ex. of requests to replay: A {from: node, nonce = X}, B {from: node, nonce = X + 1} for arbitrary X. A will be rejected, and B accepted.

Thinking on the solutions, this node1 may store max_nonce of all seen nonces from node2, and allow requests with incoming_nonce > max_nonce updating max_nonce after that (it also ignores packet loses or delays from node2).

The open question here is what to do on a node restart - nonces should be synced. I'd guess some kind of authentication happens when new connection between nodes is established, and nonces could be reset if they linked to current session. That will eliminate replies from the same node with higher nonces but previous session.

Reply:
Hmm yes you're right. We can definitely guarantee they're increasing only, but we do need a reset mechanism somehow. Looking into what's commonly done, it seems generating a random nonce is often the solution [link](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.1061.3009&rep=rep1&type=pdf) but i'm not sure how we'd sync it between 2 nodes.

Reply:
If we're only adding support for replay attacks here, and we're assuming the channel isn't "comprisable" what if we add a "ResetNonce" message which would reset the max nonce to 0. A node comes online, and broadcasts to all peers that it wants to reset it nonces with them. (Assuming some layer of node inter-authentication). In the ResetNonce, the peer would generate a random starting point, which would protect against replaying messages from (0).

Reply:
ResetNonce could be replayed too and may disrupt all communication in the quorum. My suggestion is to take into account environment/type of connections where quorum will be run and safety guarantees it provides and decide if this kind of mechanism is required on application level.